src/object_async/hello_async.cpp

```diff
@@ -0,0 +1,227 @@
+#include "hello_async.hpp"
+
+#include <exception>
+#include <stdexcept>
+#include <iostream>
+#include <map>
+
+// If this was not defined within a namespace, it would be in the global scope.
+namespace object_async {
+
+  // Custom constructor, assigns custom name passed in from Javascript world.
+  // This constructor uses member init list via the semicolon, aka "direct initialization" 
+  // which is more efficient than using assignment operators.
+  HelloObjectAsync::HelloObjectAsync(std::string name) : 
+    name_(name) {}
+
+  // Triggered from Javascript world when calling "new HelloObjectAsync()" or "new HelloObjectAsync(name)"
+  NAN_METHOD(HelloObjectAsync::New) {
+    if (info.IsConstructCall())
+    {
+        try
+        {   
+            if (info.Length() >= 1) {
+              if (info[0]->IsString()) 
+              {
+                std::string name = *v8::String::Utf8Value(info[0]->ToString());
+                auto *const self = new HelloObjectAsync(name);
+                self->Wrap(info.This());
+              }
+              else
+              {
+                return Nan::ThrowTypeError(
+                    "arg must be a string");
+              }
+            }
+            else {
+                auto *const self = new HelloObjectAsync();
+                self->Wrap(info.This());
+            } 
+
+        }
+        catch (const std::exception &ex)
+        {
+            return Nan::ThrowTypeError(ex.what());
+        }
+
+        info.GetReturnValue().Set(info.This());
+    }
+    else
+    {
+        return Nan::ThrowTypeError(
+            "Cannot call constructor as function, you need to use 'new' keyword");
+    }
+  }
+
+  /*
+  * This is an internal function used to return callback error messages instead of
+  * throwing errors.
+  * Usage:
+  *
+  * v8::Local<v8::Function> callback;
+  * CallbackError("error message", callback);
+  * return; // this is important to prevent duplicate callbacks from being fired!
+  */
+  inline void CallbackError(std::string message, v8::Local<v8::Function> callback) {
+  	v8::Local<v8::Value> argv[1] = { Nan::Error(message.c_str()) };
+  	Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+  }
+
```

Comment:
I see this same function getting repeated in multiple `.cpp` files. How about we refactor the code to move it to a new header file called `module_utils.hpp`? Then define it there, include `module_utils.hpp` in each of the `cpp` files that uses it, and that will reduce the code duplication.

Note: for fun context because of the `inline` keyword is used the compiler is doing away with this function in the binary code. When something is "inlined" it is re-arranged and optimized. So the `v8::Local<v8::Value>...` and `Nan::MakeCallback...` lines are being moved to where the code is calling `CallbackError`. This means that moving the source code into a common header will only impact how the source code looks. The final binary will probably not change at all because of how the compiler works.